### ISSUE TITLE
tracking appearing/disappearing objects across entire video

### ISSUE THREAD

**Issue by Zarxrax** [heyoeyo/muggled_sam#33]
Hi, I was evaluating the video segmentation of muggled_sam, and ran into an issue.
I tried out video_segmentaition_multiobj.py with my own example clip, which features objects which appear into the scene after the video starts. Something I noticed is that this script does not begin tracking an object until after the point prompt (for an appearing object, you will need to prompt on a frame after it appears).
The official sam2 code easily handles tracking objects on frames prior to the prompts. I was wondering if this was simply a limitation of the test script, or a limitation of muggled_sam itself?

**Comment by heyoeyo** [comment]
Hello!

The SAM models don't have any inherent notion of time before or after a prompt, they just do tracking in the sequence they're given. The multiobj script only [reads](https://github.com/heyoeyo/muggled_sam/blob/7cd3959e4728d480ecf0848c659cc839386cbd6c/simple_examples/video_segmentation_multiobj.py#L91) in the normal/forward order of the video. To do tracking before the prompt, the frames just need to be read in reverse order.

The [run_video](https://github.com/heyoeyo/muggled_sam#run-video-or-webcam) script has an example of this by toggling the `Reverse` button (just under the `Track` button, or press the `r` key).
The code for reading in reverse is something like:
```python
reversed_frame_idx_list = [5, 4, 3, 2, 1] # for example
for frame_idx in reversed_frame_idx_list:
    vcap.set(cv2.CAP_PROP_POS_FRAMES, frame_idx)
    ok_frame, frame = vcap.read()
```

If you wanted to do full forward/backward tracking, you'd want to first do a forward pass, then [reset the memory bank](https://github.com/heyoeyo/muggled_sam/blob/7cd3959e4728d480ecf0848c659cc839386cbd6c/simple_examples/video_segmentation_multiobj.py#L67) and then do a backward pass. This would be like wrapping the [multiobj loop](https://github.com/heyoeyo/muggled_sam/blob/7cd3959e4728d480ecf0848c659cc839386cbd6c/simple_examples/video_segmentation_multiobj.py#L88-L169) in an extra: `for is_reversed in (False, True):` and adding a bit of logic to switch how the frames are read & resetting the memory.

Since reversing video files is generally very slow (and open-cv doesn't guarantee frame index consistency when doing this!), it's probably better to cache frames (either in RAM or write to disk) in the forward pass and then read from this cache to handle the reverse pass. If you have very short videos (or a ton of VRAM) you can even cache/re-use the [frame encodings](https://github.com/heyoeyo/muggled_sam/blob/7cd3959e4728d480ecf0848c659cc839386cbd6c/simple_examples/video_segmentation.py#L84), in which case you can run the reverse tracking much faster (this is sort of how the original SAM2/SAM3 code does things).

**Comment by Zarxrax** [comment]
Hmmm, having to process both forward and backward for each object is unfortunate. 

I am wondering how the original sam2 manages to do it? There, lets say I put a prompt on frame 30. The object appears in the scene on frame 10. I begin propagating forward starting from frame 0. It just manages to correctly track the object on each frame that it appears, no backward pass required.

**Comment by heyoeyo** [comment]
> I am wondering how the original sam2 manages to do it?

I don't remember if SAM2 directly does a backward pass or not. It always caches a full forward pass (`predictor.init_state`) before any prompts, so it can manage forward/backward loops internally if needed. It also [flag reversals](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/sam2_video_predictor.py#L235-L239) when adding prompts, though it's not clear if this affects the memory selection later on (which occurs in [_prepare_memory_conditioned_features](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/modeling/sam2_base.py#L497) if you're interested).

It's may just use prompt encodings 'from the future' to segment objects earlier in the video. In muggled sam, this would be equivalent to [computing the prompt memory](https://github.com/heyoeyo/muggled_sam/blob/7cd3959e4728d480ecf0848c659cc839386cbd6c/simple_examples/video_segmentation_multiobj.py#L107-L108) before processing the video. In the multiobj script, for example:
```python
# Encode all prompts before processing the rest of the video
for future_idx, obj_prompt_data in prompts_per_frame_index.items():
  # Jump ahead to future frame and compute encoding
  vcap.set(cv2.CAP_PROP_POS_FRAMES, future_idx)
  ok_frame, frame = vcap.read()
  assert ok_frame, f"Invalid future frame index: {future_idx}"
  encoded_img = track_model.encode_image(frame, **imgenc_config_dict)

  # Encode/store prompt memory for each object on this frame
  for obj_key_name, obj_prompts in obj_prompt_data.items():
    init_mask, init_mem = track_model.encode_prompt_memory(encoded_img, **obj_prompts)
    memory_per_obj_dict[obj_key_name].store_prompt_result(init_mem)

# Reset video to beginning
vcap.set(cv2.CAP_PROP_POS_FRAMES, 0)
```
This would need to be done [before](https://github.com/heyoeyo/muggled_sam/blob/7cd3959e4728d480ecf0848c659cc839386cbd6c/simple_examples/video_segmentation_multiobj.py#L82) running through the whole video, and the existing [prompt memory encoding step](https://github.com/heyoeyo/muggled_sam/blob/7cd3959e4728d480ecf0848c659cc839386cbd6c/simple_examples/video_segmentation_multiobj.py#L100-L131) should be disabled/removed.

The main downside of this approach is that it'll be very error prone if the object appears later in the video, especially if there are other similar looking objects. For longer videos it could also be a lot less efficient.

**Comment by Zarxrax** [comment]
Ah, the full forward pass (predictor.init_state), is that where it loads all of the video frames into the inference state? So I guess muggled_sam doesnt do that part, which would explain part of the difference I guess.

**Comment by heyoeyo** [comment]
Ya the muggled sam version of the models are all stateless, which is one of the main differences from the originals.
The mugsam equivalent of the caching that [init_state()](https://github.com/facebookresearch/sam2/blob/2b90b9f5ceec907a1c18123530e92e794ad901a4/sam2/sam2_video_predictor.py#L42) does would be something like:
```python
state = {}
while True:
  ok_frame, frame = vcap.read()
  if not ok_frame:
    break
  state[len(state)] = track_model.encode_image(frame, **imgenc_config_dict)
```

Then the encodings can be read instead of [computing them](https://github.com/heyoeyo/muggled_sam/blob/7cd3959e4728d480ecf0848c659cc839386cbd6c/simple_examples/video_segmentation_multiobj.py#L98) in the video loop with:
```python
encoded_img = state[frame_idx]
```
This would be a lot faster for jumping around/reversing the video, but requires a lot more VRAM.

**Comment by heyoeyo** [comment]
For future reference, I've doubled checked and SAM2 doesn't do a backwards pass by default, just the [from-the-future](https://github.com/heyoeyo/muggled_sam/issues/33#issuecomment-5085222110) trick.

Here's a simple test video to demonstrate it:

<details>
<summary>Example video</summary>

https://github.com/user-attachments/assets/ad477972-de81-4cd6-a797-70edbaba5aa7
</details>

Using a prompt on the magenta circle on frame 20 (where it's on the right side of the image) results in the green circle being (incorrectly) tracked. This happens because the future prompt is a better match to the green circle when the tracking starts.

The animations below show the video on the left and the mask prediction on the right, both using a prompt on frame 20 targeting the magenta circle. The brief glitch in the SAM2 default result is actually the mask generated by the initial prompt.

SAM2 (tiny) default:

<img width="160" height="80" alt="Image" src="https://github.com/user-attachments/assets/8a8e2801-5255-44ea-921f-a242de51a998" />

Forward/backward pass:

<img width="160" height="80" alt="Image" src="https://github.com/user-attachments/assets/e71deffe-130a-487b-b30e-a6c9645a12f1" />

---

And here's the code that generates this:

<details>
<summary>SAM2 code (forward-only pass)</summary>

```python
from pathlib import Path
import cv2
import numpy as np
import torch
from sam2.build_sam import build_sam2_video_predictor

# Config
video_path = "/path/to/circles.mp4"
checkpoint = "/path/to/sam2.1_hiera_tiny.pt"
model_cfg = "/path/to/configs/sam2.1/sam2.1_hiera_t.yaml"
device, dtype = "cuda", torch.bfloat16

# Initial prompt
prompt_fidx = 20
prompt_xy = (320, 192)

with torch.inference_mode(), torch.autocast(device_type=device, dtype=dtype):
    predictor = build_sam2_video_predictor(model_cfg, checkpoint)
    inference_state = predictor.init_state(video_path)
    predictor.add_new_points_or_box(
        inference_state=inference_state,
        frame_idx=prompt_fidx,
        obj_id=1,
        points=np.array([prompt_xy], dtype=np.float32),
        labels=np.array([1], np.int32),
    )

    # Forward pass with prompts 'from the future'
    result_masks = {}
    for frame_idx, _, masks in predictor.propagate_in_video(inference_state, start_frame_idx=0):
        mask_uint8 = ((masks > 0).byte() * 255).cpu().numpy().squeeze()
        result_masks[frame_idx] = cv2.cvtColor(mask_uint8, cv2.COLOR_GRAY2BGR)

# Visualize full tracking result (have to save results to disk to avoid error!)
save_folder_path = Path(__file__).parent / f"{Path(video_path).stem}_result"
save_folder_path.mkdir(exist_ok=True)
print("", "", "Saving results to:", save_folder_path, sep="\n")
vcap = cv2.VideoCapture(video_path)
for fidx in range(1_000_000):
    ok_frame, frame = vcap.read()
    if not ok_frame:
        break
    result = result_masks.get(fidx, None)
    if result is None:
        result = frame * 0
    cv2.imwrite(save_folder_path / f"{fidx:0>5}.png", np.hstack((frame, result)))
    # cv2.imshow("Result", result) # <- This gives an error! SAM2 seems to be using it...
    # cv2.waitKey(100)
vcap.release()
```
</details>


<details>
<summary>Forward/backward code</summary>

```python
from time import perf_counter
import cv2
import numpy as np
import torch
from muggled_sam.make_sam import make_sam_from_state_dict

# Config
video_path = "/path/to/circles.mp4"
model_path = "/path/to/sam2.1_hiera_tiny.pt"
device, dtype = "cuda", torch.bfloat16
imgenc_config = {"max_side_length": None, "use_square_sizing": True}

# Initial prompt
prompt_frame_idx = 20
prompt_xy = (0.835, 0.5)

# Cache frames from video (so we can run backwards quickly)
print("Caching video frames...")
vcap = cv2.VideoCapture(video_path)
frame_cache = []
while True:
    ok_frame, frame = vcap.read()
    if not ok_frame:
        break
    frame_cache.append(frame)
vcap.release()

# Get initial prompt memory
print("Loading model...")
track_model = make_sam_from_state_dict(model_path).get_tracking_context().to(device, dtype)
init_frame = frame_cache[prompt_frame_idx]
init_encimg = track_model.encode_image(init_frame, **imgenc_config)
_, init_mem = track_model.encode_prompt_memory(init_encimg, None, [prompt_xy], None)

# Forward/backward pass
print("Processing frames...")
t1 = perf_counter()
result_masks = {}
all_idx_list = list(range(0, len(frame_cache)))
back_iter, forward_iter = all_idx_list[0:prompt_frame_idx], all_idx_list[prompt_frame_idx:]
for idx_iter in (forward_iter, reversed(back_iter)):
    p_mems, f_mems = [init_mem], [] # Reset 'memory bank'
    for fidx in idx_iter:
        frame = frame_cache[fidx]
        enc_img = track_model.encode_image(frame, **imgenc_config)
        mask, _, ptr, score = track_model.step_video_masking(enc_img, p_mems, f_mems[-6:])
        if score < 0:
            break # or 'continue' to keep searching for object
        f_mems.append(track_model.encode_frame_memory(enc_img, mask, ptr, score))
        mask_scaled = torch.nn.functional.interpolate(mask, size=frame.shape[0:2], mode="bilinear")
        mask_uint8 = ((mask_scaled > 0).byte() * 255).cpu().numpy().squeeze()
        result_masks[fidx] = cv2.cvtColor(mask_uint8, cv2.COLOR_GRAY2BGR)
t2 = perf_counter()
print(f"  -> Took {round(1000 * (t2 - t1))} ms")

# Visualize full tracking result
for fidx, frame in enumerate(frame_cache):
    result = result_masks.get(fidx, None)
    if result is None:
        result = frame * 0
    cv2.imshow("Result", np.hstack((frame, result)))
    cv2.waitKey(100)
cv2.destroyAllWindows()
```
</details>

**Comment by Zarxrax** [comment]
Ah, very interesting. Thanks for the insight. I'll keep this video around for future testing.

**Comment by heyoeyo** [comment]
For the sake of completeness, I figured it was worth double-checking this for SAM3 as well. At least when tracking from prompts (i.e. not detections) SAM3 also does 'from the future' tracking.
Here's a minimal code example (based on the [example notebook](https://github.com/facebookresearch/sam3/blob/main/examples/sam3_for_sam2_video_task_example.ipynb), running on an older commit eg. [11dec29](https://github.com/facebookresearch/sam3/tree/11dec2936de97f2857c1f76b66d982d5a001155d)):
<details>
<summary>SAM3.0 code (forward-only pass)</summary>

```python
from pathlib import Path
import torch
import cv2
import numpy as np
from sam3.model_builder import build_sam3_video_model

# Config
video_path = "/path/to/circles.mp4"
checkpoint = "/path/to/sam3.pt"
bpe_path = "/path/to/sam3/assets/bpe_simple_vocab_16e6.txt.gz"

# Initial prompt
prompt_frame_index = 20
prompt_xy = (0.835, 0.5)

# Set up model & 'future' prompt
sam3_model = build_sam3_video_model(checkpoint, bpe_path=bpe_path)
predictor = sam3_model.tracker
predictor.backbone = sam3_model.detector.backbone
inference_state = predictor.init_state(video_path=video_path)
predictor.clear_all_points_in_video(inference_state)
_, out_obj_ids, low_res_masks, video_res_masks = predictor.add_new_points(
    inference_state=inference_state,
    frame_idx=prompt_frame_index,
    points=torch.tensor([prompt_xy], dtype=torch.float32),
    labels=torch.tensor([1], dtype=torch.int32),
    obj_id=1,
)

# Forward pass only
result_masks = {}
for frame_idx, _, _, video_res_masks, _ in predictor.propagate_in_video(
    inference_state,
    start_frame_idx=0,
    max_frame_num_to_track=None,
    reverse=False,
    propagate_preflight=True,
):
    mask_uint8 = ((video_res_masks > 0).byte() * 255).cpu().numpy().squeeze()
    result_masks[frame_idx] = cv2.cvtColor(mask_uint8, cv2.COLOR_GRAY2BGR)

# Visualize full tracking result (have to save results to disk to avoid error!)
save_folder_path = Path(__file__).parent / f"{Path(video_path).stem}_result"
save_folder_path.mkdir(exist_ok=True)
print("", "", "Saving results to:", save_folder_path, sep="\n")
vcap = cv2.VideoCapture(video_path)
for fidx in range(1_000_000):
    ok_frame, frame = vcap.read()
    if not ok_frame:
        break
    result = result_masks.get(fidx, None)
    if result is None:
        result = frame * 0
    combined = np.hstack((frame, result))
    cv2.imwrite(save_folder_path / f"{fidx:0>5}.png", combined)
    # cv2.imshow("Result", combined)  # <- This gives an error! SAM3 seems to be using it...
    # cv2.waitKey(100)
vcap.release()
```
</details>

---

Interestingly the v3.1 model gives a correct looking result for the `circles.mp4` video above even with a future prompt, as it's better at matching the object appearance (i.e. less reliant on positioning) from the prompt.
So here's a slight variation to test the forward-only tracking for all model versions:

https://github.com/user-attachments/assets/e98e7653-616a-4d0d-8a4a-de21c23a99c4

As an example, a prompt on the magenta _square_ later in the video (frame 60) leads the 3.1 forward-only pass to incorrectly segment the magenta _circle_ at first, and it's also late tracking the square when it does appear:

SAM3.1 Forward-only

<img width="160" height="80" alt="Image" src="https://github.com/user-attachments/assets/5d6aa9c7-f0c9-45fd-8434-a43ad8eb380f" />

By comparison, the forward/back pass cleanly tracks only the square as expected. The forward/back pass is also faster in this case (4s vs. 11s) since it doesn't need to process the whole video to search for the object:

SAM3.1 Forward/backward

<img width="160" height="80" alt="Image" src="https://github.com/user-attachments/assets/b039d3a0-939c-42a3-8b64-c798a8640df9" />

Here's the test code for the v3.1 forward-only pass (also works for v2/v3.0 models):

<details>
<summary>SAM3.1 code (forward-only pass, MuggledSAM version)</summary>

```python
from time import perf_counter
import cv2
import numpy as np
import torch
from muggled_sam.make_sam import make_sam_from_state_dict

# Config
video_path = "/path/to/circles_diagonal.mp4"
model_path = "/path/to/sam3.1_multiplex.pt"
device, dtype = "cuda", torch.bfloat16
imgenc_config = {"max_side_length": None, "use_square_sizing": True}

# Initial prompt
prompt_frame_idx = 60
prompt_xy = (0.5, 0.4)

# Cache frames from video
print("Caching video frames...")
vcap = cv2.VideoCapture(video_path)
frame_cache = []
while True:
    ok_frame, frame = vcap.read()
    if not ok_frame:
        break
    frame_cache.append(frame)
vcap.release()

# Get 'future' prompt memory
print("Loading model...")
track_model = make_sam_from_state_dict(model_path).get_tracking_context().to(device, dtype)
init_frame = frame_cache[prompt_frame_idx]
init_encimg = track_model.encode_image(init_frame, **imgenc_config)
_, init_mem = track_model.encode_prompt_memory(init_encimg, None, [prompt_xy], None)
p_mems, f_mems = [init_mem], []

# Forward-only pass
print("Processing frames...")
t1 = perf_counter()
result_masks = {}
for fidx, frame in enumerate(frame_cache):
    enc_img = track_model.encode_image(frame, **imgenc_config)
    mask, _, ptr, score = track_model.step_video_masking(enc_img, p_mems, f_mems[-6:])
    if score < 0:
        continue # Can't use 'break' here, since the object may not exist on early frames!
    f_mems.append(track_model.encode_frame_memory(enc_img, mask, ptr, score))
    mask_scaled = torch.nn.functional.interpolate(mask, size=frame.shape[0:2], mode="bilinear")
    mask_uint8 = ((mask_scaled > 0).byte() * 255).cpu().numpy().squeeze()
    result_masks[fidx] = cv2.cvtColor(mask_uint8, cv2.COLOR_GRAY2BGR)
t2 = perf_counter()
print(f"  -> Took {round(1000 * (t2 - t1))} ms")

# Visualize full tracking result
for fidx, frame in enumerate(frame_cache):
    result = result_masks.get(fidx, None)
    if result is None:
        result = frame * 0
    cv2.imshow("Result", np.hstack((frame, result)))
    cv2.waitKey(100)
cv2.destroyAllWindows()
```
</details>

The v3.1 forward/backward code is the same as in the [previous post](https://github.com/heyoeyo/muggled_sam/issues/33#issuecomment-5097297962), except the prompt xy is (0.5, 0.4) on frame 60 for this video.
